improver_tests/utilities/cube_extraction/test_ExtractPressureLevel.py

```diff
@@ -102,34 +102,137 @@ def cube_shape_check_without_realizations(pressure_slice_cube):
     assert pressure_slice_cube.shape == (3, 2)
 
 
+@pytest.mark.parametrize("least_significant_digit", (0, None))
+@pytest.mark.parametrize("reverse_pressure", (False, True))
+@pytest.mark.parametrize(
+    "special_value", (None, np.nan, True, np.inf, (np.nan, np.nan))
+)
 @pytest.mark.parametrize("with_realization", (True, False))
 @pytest.mark.parametrize(
-    "temperature,expected",
+    "temperature,expected_p_index",
     (
-        (280, 80000),  # Exactly matches a pressure value
-        (277, 75000),  # Half way between pressure values
-        (301, 100000),  # Temperature above max snaps to max
-        (244, 30000),  # Temperature below min snaps to min
+        (280, 2),  # Exactly matches a pressure value
+        (277, 2.5),  # Half way between pressure values
+        (301, 0),  # Temperature above max snaps to pressure at max
+        (244, 7),  # Temperature below min snaps to pressure at min
     ),
 )
 def test_basic(
-    temperature, temperature_on_pressure_levels, expected, with_realization,
+    temperature,
+    temperature_on_pressure_levels,
+    expected_p_index,
+    with_realization,
+    special_value,
+    reverse_pressure,
+    least_significant_digit,
 ):
     """Tests the ExtractPressureLevel plugin with values for temperature and
     temperature on pressure levels to check for expected result.
+    Tests behaviour when temperature and/or pressure increase or decrease along
+    the pressure axis.
+    Tests behaviour with different special values in the temperature data.
+    Tests behaviour with and without a realization coordinate.
     Also checks the metadata of the output cube"""
+    special_value_index = 0
+    positive_correlation = True
+    if reverse_pressure:
+        # Flip the pressure coordinate for this test. We also swap which end the
+        # special value goes, so we can test _one_way_fill in both modes.
+        temperature_on_pressure_levels.coord(
+            "pressure"
+        ).points = temperature_on_pressure_levels.coord("pressure").points[::-1]
+        special_value_index = -1
+        positive_correlation = False
+    expected = np.interp(
+        expected_p_index,
+        range(len(temperature_on_pressure_levels.coord("pressure").points)),
+        temperature_on_pressure_levels.coord("pressure").points,
+    )
     expected_data = np.full_like(
         temperature_on_pressure_levels.data[:, 0, ...], expected
     )
+
+    if special_value is True:
+        # This is a proxy for setting a mask=True entry
+        temperature_on_pressure_levels.data = np.ma.MaskedArray(
+            temperature_on_pressure_levels.data, mask=False
+        )
+        temperature_on_pressure_levels.data.mask[
+            0, special_value_index, 0, 0
+        ] = special_value
+    elif special_value is None:
+        pass
+    else:
+        temperature_on_pressure_levels.data = temperature_on_pressure_levels.data.copy()
+        if isinstance(special_value, float):
```

Comment:
This is all getting pretty unlovely, using a float instance check to separate out your tuple. Yuck.

Reply:
I don't like this either, so it now checks for an `Iterable`, which inverts the if/else blocks, but works and is less ugly.